### ISSUE TITLE
Mix precision leads to nan loss

### ISSUE THREAD

**Issue by ZhenShuo2021** [keras-team/keras#16552]
**System information**.
- Have I written custom code (as opposed to using a stock example script provided in Keras): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 21h2 (19044, 1706)
- TensorFlow installed from (source or binary): pip install
- TensorFlow version (use command below): tf2.8
- Python version: 3.9.7
- Bazel version (if compiling from source): unknown
- GPU model and memory: RTX3070/8GB
- Exact command to reproduce: No

**Describe the problem**.

Loss is nan when using mix precision API even though use (get_scaled_loss, get_unscaled_gradients) or scale loss manually.

- Model works well in float32.
- Mix precision API with [tensorflow guide](https://www.tensorflow.org/guide/mixed_precision#training_the_model_with_a_custom_training_loop) works well in my computer.
- My model input start from Cond2D -> ... -> Resblock -> ... -> Dense -> SoftMax, it's to solve a classification problem.

**Describe the current behavior**.

- Inf loss for optimizer.get_scaled_loss, or
- Nan gradient after scaled loss

**Describe the expected behavior**.

Loss and gradient scaled like tensorflow guide.

- Do you want to contribute a PR? (no):
- If yes, please read [this page](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md) for instructions
- Briefly describe your candidate solution(if contributing):

**Standalone code to reproduce the issue**.

```shell
import os
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3'
import numpy as np
import tensorflow as tf
from utils._history import fig
from utils.network_float16 import MyModel, CosineAnnealingScheduler, CCE, Scale_Gradient
from utils.system import args, Yie, _Plot, _Predict, MyTrain
tf.keras.mixed_precision.set_global_policy('mixed_float16')

test_SNR = np.arange(0, 22, 2) if args.fading_type == 1 else np.arange(10, 32, 2)
a,b = MyTrain(args.K, args.m, args.Nd, args.B, args.p, args.Ka, args.snr, args.tap, args.fading_type)   # training data generation

train_dataset = tf.data.Dataset.from_tensor_slices((a, b)) 
train_dataset = train_dataset.shuffle(buffer_size=2048).batch(args.B)
optimizer = tf.keras.optimizers.Adam(learning_rate=5e-4)
optimizer = tf.keras.mixed_precision.LossScaleOptimizer(optimizer)
loss_metric = tf.keras.metrics.CategoricalCrossentropy()
acc = tf.keras.metrics.Accuracy()

model = MyModel(args)
model = model.Model(is_summary=0, is_plot_model=1)   # compile model


@tf.function
def train_step(x_train, y_train):
    scale=2048
    with tf.GradientTape() as tape:
        y_pred = model(x_train,training = True)
        loss = tf.keras.losses.categorical_crossentropy(y_train, y_pred)
        loss *= scale
    grads = tape.gradient(loss, model.trainable_weights)
    grads = [x*scale for x in grads]
    optimizer.apply_gradients(zip(grads, model.trainable_weights))
    return loss


def decision(y_pred):
    y_pred = np.squeeze(np.argsort(-y_pred)[:, :args.Ka])
    temp = np.zeros([args.B, args.K], dtype='int')
    for j in range(args.Ka):
        temp[np.arange(args.B), y_pred[:, j]] = 1
    return temp

for epoch in range(args.epoch):
    epoch_time = time.time()
    for step, (x_train, y_train) in enumerate(train_dataset):
        loss = train_step(x_train, y_train)
    y_pred = model(x_train)
    y_pred = decision(y_pred)
    acc.update_state(y_pred, y_train)
    print("Epoch %d - %ds - loss: %.4f - acc: %4f" % (epoch+1, time.time()-epoch_time, loss_record[epoch], acc.result().numpy()))

_Plot.Simulation(args, _Predict(args, model, test_SNR), test_SNR, fig, save_txt=1)
##### Loss is nan from first epoch #####



##### Next, I tried two types of train_step, and the 'Revelant log output' is the print output #####
##### First I set the size of x_train to batch size, so the network will only update weight ones #####
##### Compile the model and run the following code #####
model = MyModel(args)
model = model.Model(is_summary=0, is_plot_model=1)   # compile model
# type 1 (follow tensorflow guide)

for scale in range(1,3000,30):   # to check which scale is the best
    with tf.GradientTape() as tape:
        y_pred = model(x_train,training = True)
        loss = tf.keras.losses.categorical_crossentropy(y_train, y_pred)
        print(tf.reduce_mean(loss))    # see if CCE loss works normally
        loss = optimizer.get_scaled_loss(loss)
        print(tf.reduce_mean(loss))   # see if get_scaled_loss works normally
    grads = tape.gradient(loss, model.trainable_weights)
    grads = optimizer.get_unscaled_gradients(grads)
    optimizer.apply_gradients(zip(grads, model.trainable_weights))
# -----------------------------------------------------------------
# type 2 (scale loss manually)

for scale in range(1,3000,30):   # to check which scale is the best
    with tf.GradientTape() as tape:
        y_pred = model(x_train,training = True)
        loss = tf.keras.losses.categorical_crossentropy(y_train, y_pred)
        loss *= scale
        print(tf.reduce_mean(loss))   # see if loss scaled properly
    grads = tape.gradient(loss, model.trainable_weights)
    print(tf.reduce_mean(grads[0]))  #  _ in grads[ _ ] can be arbitrary number (layer)
    grads = [x*scale for x in grads]
    optimizer.apply_gradients(zip(grads, model.trainable_weights))

```

**Source code / logs**.

```shell
# type 1

tf.Tensor(9.516, shape=(), dtype=float16)   # unscaled loss
tf.Tensor(inf, shape=(), dtype=float16)     # scaled loss
tf.Tensor(9.516, shape=(), dtype=float16)   # unscaled loss
tf.Tensor(inf, shape=(), dtype=float16)     # scaled loss
...
... # same for other scales
# -------------------------------------------------------------------------
# type 2

tf.Tensor(9.516, shape=(), dtype=float16)    # scaled loss
tf.Tensor(nan, shape=(), dtype=float32 )     # gradients
tf.Tensor(961.0, shape=(), dtype=float16)    # scaled loss
tf.Tensor(nan, shape=(), dtype=float32)      # gradients
tf.Tensor(1913.0, shape=(), dtype=float16)   # scaled loss
tf.Tensor(nan, shape=(), dtype=float32)      # gradients
...
...  # gradients are all nan for all scales

##### All gradients are nan from first step, even w/wo kernel initializer (random normal initializer) #####
```

**Comment by tilakrayal** [comment]
@ZhenShuo2021 ,
While trying to execute the mentioned code, I am facing issue to import the required `utils` for the code which are not relatable to tensorflow. Could you provide the required dependencies for the code to be executed.It helps to debug the issue.Thanks!
